source/postprocess/visualization/compositional_vector.cc

```diff
@@ -0,0 +1,202 @@
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <aspect/postprocess/visualization/compositional_vector.h>
+#include <aspect/simulator_access.h>
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      CompositionalVector<dim>::
+      CompositionalVector ()
+        :
+        DataPostprocessor<dim> ()
+      {}
+
+
+      template <int dim>
+      std::vector<std::string>
+      CompositionalVector<dim>::
+      get_names () const
+      {
+        std::vector<std::string> solution_names;
+        for (unsigned int i=0; i<vector_names.size(); ++i)
+          for (unsigned int j=0; j<dim; ++j)
+            solution_names.push_back(vector_names[i]);
+        return solution_names;
+      }
+
+
+      template <int dim>
+      std::vector<DataComponentInterpretation::DataComponentInterpretation>
+      CompositionalVector<dim>::
+      get_data_component_interpretation () const
+      {
+        std::vector<DataComponentInterpretation::DataComponentInterpretation> interpretation;
+        for (unsigned int i=0; i<vector_names.size()*dim; ++i)
+          interpretation.push_back (DataComponentInterpretation::component_is_part_of_vector);
+
+        return interpretation;
+      }
+
+
+      template <int dim>
+      UpdateFlags
+      CompositionalVector<dim>::
+      get_needed_update_flags () const
+      {
+        return update_values;
+      }
+
+
+      template <int dim>
+      void
+      CompositionalVector<dim>::
+      compute_derived_quantities_vector (const std::vector<Vector<double> >              &uh,
+                                         const std::vector<std::vector<Tensor<1,dim> > > &,
+                                         const std::vector<std::vector<Tensor<2,dim> > > &,
+                                         const std::vector<Point<dim> > &,
+                                         const std::vector<Point<dim> > &,
+                                         std::vector<Vector<double> >                    &computed_quantities) const
+      {
+        const unsigned int n_quadrature_points = uh.size();
+        Assert (computed_quantities.size() == n_quadrature_points, ExcInternalError ());
+        Assert (uh[0].size() == this->introspection().n_components, ExcInternalError ());
+
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            std::vector<double> composition(this->n_compositional_fields());
+
+            for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+              composition[c] = uh[q][this->introspection().component_indices.compositional_fields[c]];
+
+            for (unsigned int i=0; i<vector_names.size(); ++i)
+              for (unsigned int j=0; j<dim; ++j)
+                computed_quantities[q][i*dim+j] = composition[sets[i][j]];
```

Comment:
This seems inefficient. You first copy *all* components from `uh[q]` into the `composition` array, and then you subset this field into `computed_quantities`. I think you could do this in one step without the copy operation.

If you think it's clearer the way you have it right now, please pull the declaration of the `composition` array out of the loop.